gourmet/gtk/recipe.py

```diff
@@ -0,0 +1,787 @@
+    # Reject edits if the edited row is an ingredient group name
+    iter = model.get_iter_from_string(path)
+    if not isinstance(model.get_value(iter, 4), _IngredientGroup):
+        value = model.get_value(iter, column)
+        model.set_value(iter, column, not value)
+
+
+def _add_group(parent, ingredients):
+    dialog = _Gtk.Dialog(
+        title = _('New Ingredient Group'),
+        transient_for = parent,
+        destroy_with_parent = True
+    )
+
+    main = _Gtk.Box(
+        orientation = _Gtk.Orientation.HORIZONTAL,
+        margin_bottom = 6,
+        spacing = 6
+    )
+    main.pack_start(_Gtk.Label(label=_('Group Name:')), False, False, 0)
+
+    entry = _Gtk.Entry()
+    def enter_group(_):
+        if entry.get_text():
+            dialog.response(_Gtk.ResponseType.OK)
+    entry.connect('activate', enter_group)
+    main.pack_start(entry, True, True, 0)
+
+    content = dialog.get_content_area()
+    content.set_property('margin', 6)
+    content.pack_start(main, True, True, 0)
+
+    dialog.add_button('Cancel', _Gtk.ResponseType.CANCEL)
+    dialog.add_button('Ok', _Gtk.ResponseType.OK)
+
+    dialog.show_all()
+    response = dialog.run()
+
+    if _Gtk.ResponseType.OK == response:
+        group_name = entry.get_text()
+        group = _IngredientGroup()
+        ingredient_iter = ingredients.ingredients_tree.get_model().append(
+                None, [None, None, group_name, None, group, None])
+        group.iter = ingredients.group_list.append([group_name, ingredient_iter])
+        ingredients.group_combobox.set_active_iter(group.iter)
+        ingredients.group_combobox.set_sensitive(True)
+
+    dialog.destroy()
+
+
+def _add_ingredient(ingredients, rec_gui):
+    text = ingredients.ingredient_entry.get_text().strip()
+    if text:
+        result = rec_gui.rd.parse_ingredient(text)
+        if 'rangeamount' in result:
+            result['amount'] = rec_gui.rd.rd._format_amount_string_from_amount(
+                (result['amount'], result['rangeamount'])
+            )
+            del result['rangeamount']
+        elif 'amount' in result:
+            result['amount'] = _convert.float_to_frac(result['amount'])
+        recipe_id = _check_if_recipe(rec_gui.rd, result['item'])
+        group_iter = ingredients.group_combobox.get_active_iter()
+        parent = ingredients.group_combobox.get_model().get_value(group_iter, 1)
+        ingredients.ingredients_tree.get_model().append(parent, [
+            result.get('amount', None),
+            result.get('unit', None),
+            result['item'],
+            result.get('optional', False),
+            None,
+            recipe_id,
+        ])
+        if parent:
+            ingredients.ingredients_tree.expand_row(
+                ingredients.ingredients_tree.get_model().get_path(parent),
+                False
+            )
+    ingredients.ingredient_entry.set_text('')
+    ingredients.ingredient_entry.grab_focus()
+
+
+def _check_if_recipe(db, name):
```

Comment:
It's debatable whether a recipe title can be used as a unique ID. 
It's certainly interesting, but could break user's workflow.  
Have you tried to have a database that already contains 2 recipes with the same title? I can think of a number of scenarios where it might be the case (eg. imports from backups, from other applications).  

I wonder what would happen if a user had their database with duplicates and would update to a Gourmet with this.

Reply:
There's no enforcement of this from either the UI or the database, but recipe names _should_ be unique, even if the user should want to enter multiple recipes for the same dish (e.g., multiple recipes for "brownies"), otherwise they have no way to distinguish different recipes (other than reading the recipe). Or, alternatively, enforcing some unique combination of recipe name and, say, source. This does require adding a unique key to the database and adding an upgrade step that ensures the key is satisfied. While this might be a short-term or in-the-moment inconvenience for users, in the long run it should reduce duplication, produce a better-quality recipe database, and enable more powerful features. Ideally, this change should happen in conjunction with a refactoring of the database.

Reply:
I have noticed, in export data, a line for each recipe (example):

	<recipe id="1247">

Is that not a unique identifier stored in the database?

If it doesn't exist already, I think that adding a unique key to the database is a good idea, even if it requires an upgrade step for users. I also agree that names should be unique, but don't think that is a good unique key. A problem may arise if users' current database has duplicated names. It would be a really good idea to give the user a warning if they are attempting to add/edit a recipe where the name would be duplicated - and disallow save until the name is unique.

Reply:
There is a unique integer key in the database which is used as the primary key, which we will certainly want to keep as [a surrogate key](https://en.wikipedia.org/wiki/Surrogate_key), however there's no enforcement of [the natural key](https://en.wikipedia.org/wiki/Natural_key), which is how users distinguish different recipes. I would suggest using the name of the recipe as the natural key and putting a uniqueness constraint on it. Right now, there's nothing preventing a user from entering different recipes with the same name, and in fact Gourmet currently bundles a plugin to find duplicate recipes. Now certainly we want users to be able to enter multiple recipes for something, but we should make sure the user is able to somehow distinguish the different recipes (and probably alert them that they already have a similar recipe), and the easiest way would be to require a unique recipe name. As an example from the world of cocktails, notice how many different Cosmopolitan cocktail recipes are listed, but they're all distinguished by name: https://www.diffordsguide.com/search?q=cosmopolitan

The fact that duplicate recipe names have been allowed until now means we will need some sort of database upgrade step to sort this out. It can either specifically prompt the user for unique names, which may not be practical if there are a large number of duplicates in the database, or we can algorithmically make the names unique (which could be as simple as adding "1", "2", etc. to the end of the name).

Reply:
Ah, I understand what you're saying, now. To be honest, I've never understood why duplicate names were allowed. Duplicated names, as you say, make it more difficult to differentiate between similar recipes.

I lean toward doing it algorithmically to make the upgrade transition as quick as possible.

Maybe provide the user with a list of renamed recipes (old and new names) - either a text file or something they can copy/paste into a document to save for later use. If not providing the data, as described, make the rename add some text to the new name that the user can search for. For example, two recipes named Brownies could be renamed as 'Brownies renamed 1' and 'Brownies renamed 2'. Either method would allow the user to go in, at their leisure, and fix any names that were previously duplicated.